File: leetcode/leetcode-148.py
# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:
    def sortList(self, head: ListNode) -> ListNode:
        n = 0
        u = head
        while u:
            u = u.next
            n += 1
        def merge(n, u):
            if n <= 1:
                return u
            if n == 2:
                p = u.next
                if p.val > u.val:
                    p.next = u
                    u.next = None
                    return p
                return u
            k = 0
            t = u
            while k <= n//2:
                k += 1
                t = t.next
            k = merge(n//2, u)
            p = merge(n-n//2, t)
            q = ListNode(1)
            ans = q
            while k is not None and p is not None:
                if k.val < p.val:
                    q.next = k
                    k = k.next
                else:
                    q.next = p
                    p = p.next
                q = q.next
            while k is not None:
                q.next = k
                k = k.next
                q = q.next
            while p is not None:
                q.next = p
                p = p.next
                q = q.next
            return ans.next
        return merge(n, head)

File: leetcode/leetcode-148.cpp
/*
 *@File Name          :leetcode-148.cpp
 *@Author             :cc
 *@Version            :1.0.0
 *@Date               :2022/3/7 19:58
 *@Description        :
 *@Function List      :
 *@History            :
 */


struct ListNode {
    int val;
    ListNode *next;

    ListNode() : val(0), next(nullptr) {}

    ListNode(int x) : val(x), next(nullptr) {}

    ListNode(int x, ListNode *next) : val(x), next(next) {}
};


class Solution {
public:
    ListNode *sortList(ListNode *head) {
        return sortList(head, nullptr);
    }

    ListNode *sortList(ListNode *head, ListNode *tail) {
        if (head == nullptr) return head;
        if (head->next == tail) {
            head->next = nullptr;
            return head;
        }
        ListNode *fast = head, *slow = head;
        while (fast != tail) {
            fast = fast->next;
            slow = slow->next;
            if (fast != tail)
                fast = fast->next;
        }
        return mergeList(sortList(head, slow), sortList(slow, tail));
    }

    ListNode *mergeList(ListNode *l1, ListNode *l2) {
        ListNode *h = new ListNode();
        ListNode *ans = h;
        ListNode *t1 = l1, *t2 = l2;
        while (t1 != nullptr && t2 != nullptr) {
            if (t1->val <= t2->val) {
                h->next = t1;
                t1 = t1->next;
            } else {
                h->next = t2;
                t2 = t2->next;
            }
            h = h->next;
        }
        if (t1) h->next = t1;
        if (t2) h->next = t2;
        return ans->next;
    }
};